### ISSUE TITLE
ConnectionError when using DockerRun with `host_config = {"extra_hosts": {host: ip}}`

### ISSUE THREAD

**Issue by VincentAntoine** [PrefectHQ/prefect#5444]
## Description
In [`DockerRun`](https://docs.prefect.io/api/latest/run_configs.html#dockerrun) it is possible to pass `docker run` options as a `host_config` argument. This argument expects a dictionary and its content is used as runtime arguments by the Docker Agent. Among [other things](https://docker-py.readthedocs.io/en/stable/api.html#docker.api.container.ContainerApiMixin.create_host_config), it is possible to include an `extra_hosts` entry, which is the equivalent of adding a line to the `/etc/hosts` files in the container, mapping a host and its corresponding IP address to which it will be resolved.

An exemple of a `DockerRun` instance adding a mapping that will be available inside the container running the flow might be :

```python
DockerRun(
    host_config: {
        "extra_hosts": {"myapp.internal.api": "12.34.56.78"}
    })
```

Unfortunately, using this breaks the communication with Prefect server and leads to :

## Expected Behavior
When using the `extra_hosts` to add a host:IP mapping in the container, the communication with the server should still work.

## Reproduction

```python
from prefect import Flow, task
from prefect.run_configs.docker import DockerRun

@task
def do_it():
    logger = prefect.context.get("logger")
    logger.info("Done!")

with Flow("my_flow") as flow:
    do_it()

flow.run_config = DockerRun(
    host_config: {
        "extra_hosts": {"myapp.internal.api": "12.34.56.78"}
    }
)

flow.register()
```

## Environment
```
$ prefect diagnostics
{
  "config_overrides": {},
  "env_vars": [],
  "system_information": {
    "platform": "Linux-5.13.0-28-generic-x86_64-with-glibc2.29",
    "prefect_backend": "server",
    "prefect_version": "0.15.13",
    "python_version": "3.8.7"
  }
}
```

## Proposed solution

[Here](https://github.com/PrefectHQ/prefect/blob/fc9c8f07555d403093ac18f751dad07c435b51db/src/prefect/agent/docker/agent.py#L427-L429) :
```python
class DockerAgent(Agent):
    def deploy_flow(self, flow_run: GraphQLResult) -> str:
        ...
        run_config = self._get_run_config(flow_run, DockerRun)
        ...
        host_config["extra_hosts"] = {"host.docker.internal": "host-gateway"}
        ...
        host_config.update(run_config.host_config)
```

The `host_config` (the defaults from the Docker Agent) is updated with the contents of the run_config.host_config (of the DockerRun). This is OK if the run_config.host_config does not contain an `extra_hosts` entry, but if it does, then the `{"host.docker.internal": "host-gateway"}` entry is lost, which causes the error.

Two options to solve this :
- let users solve it by adding an additional `{"host.docker.internal": "host-gateway"}` entry to the host_config of the DockerRun (which I'm doing but took me some time to find what was happening, and was not very intuitive)
- modify the code in the Docker Agent like this :

```python
class DockerAgent(Agent):
    def deploy_flow(self, flow_run: GraphQLResult) -> str:
        ...
        run_config = self._get_run_config(flow_run, DockerRun)
        ...
        host_config["extra_hosts"] = {"host.docker.internal": "host-gateway"}
        ...
        if "extra_hosts" in run_config.host_config:
             run_config_extra_hosts = run_config.host_config.pop("extra_hosts")
        host_config.update(run_config.host_config)
        host_config["extra_hosts"].update(run_config_extra_hosts)
```

**Comment by madkinsz** [comment]
Hey @VincentAntoine, thanks for the well written issue! This is a little tricky, because when you provide `extra_hosts`, you are opting into management of this setting. If we always merge the internal host in, we've blocked you disabling that behavior without adding another flag. We can detect when it is missing and display a warning when it is likely to cause connection problems (when using linux and the server is on localhost) so that the solution is more apparent. What do you think?

**Comment by VincentAntoine** [comment]
Hi @madkinsz, yes, that would be very helpful.

However, I'm probably missing something but I'm not sure to understand how it could hurt to merge the internal host :
- case A : if the user **does** need a custom "host.docker.internal" in the container that runs the flow, she can still add it to the DockerRun extra_hosts and it will take precedence over the "host.docker.internal" of the defaults
- case B : if the user **does not** need a custom "host.docker.internal" host in the container that runs flows, the default "host.docker.internal" will be used (and most probably is what they need - if not, then they are in fact in case A)

Again I might be missing something, and a warning would be helpful already, but I was surprised that by adding a host to map an API endpoint I broke the communication with Docker, so if this makes sense to you then it might be worth merging the default host.

**Comment by madkinsz** [comment]
You are correct on your assumptions there, but what if you do not want `host.docker.internal` to be set at all?

It may be worth merging it in all the time anyway, it's just likely we'll get a complaint sometime in the future :) We can also see if docker complains if `host.docker.internal` is mapped to `""` or `None` which would allow users to "unset" it by specifying it.
